File: cpp/lib/testcontainers/system/ip/Ipv4Addr.cpp
#include <rust/cxx.h>
#include <rust_tc_bridge/lib.h>

#include "testcontainers/system/ip/Ipv4Addr.hpp"

#include "details/BoxHelper.hpp"

namespace testcontainers {

Ipv4Addr::Ipv4Addr(RsIpv4Addr *addr) noexcept
    : rimpl_(addr, [](RsIpv4Addr *a) { ::rs_ipv4_addr_destroy(details::box_from_raw(a)); }) {}

Ipv4Addr::Ipv4Addr(std::uint8_t a, std::uint8_t b, std::uint8_t c, std::uint8_t d) noexcept
    : Ipv4Addr(::rs_ipv4_addr_new(a, b, c, d).into_raw()) {}

Ipv4Addr::Ipv4Addr(Ipv4Addr &&other) noexcept = default;

Ipv4Addr &Ipv4Addr::operator=(Ipv4Addr &&other) noexcept = default;

Ipv4Addr::~Ipv4Addr() noexcept = default;

Ipv4Addr Ipv4Addr::from_bits(std::uint32_t bits) noexcept {
  return Ipv4Addr(::rs_ipv4_addr_from_bits(bits).into_raw());
}

Ipv4Addr Ipv4Addr::localhost() noexcept { return Ipv4Addr(::rs_ipv4_addr_localhost().into_raw()); }

Ipv4Addr Ipv4Addr::unspecified() noexcept { return Ipv4Addr(::rs_ipv4_addr_unspecified().into_raw()); }

Ipv4Addr Ipv4Addr::broadcast() noexcept { return Ipv4Addr(::rs_ipv4_addr_broadcast().into_raw()); }

Ipv4Addr Ipv4Addr::from_octets(const std::array<std::uint8_t, 4> &octets) noexcept {
  return Ipv4Addr(::rs_ipv4_addr_from_octets(octets).into_raw());
}

std::uint32_t Ipv4Addr::to_bits() const noexcept { return rimpl_->rs_ipv4_addr_to_bits(); }

std::array<std::uint8_t, 4> Ipv4Addr::octets() const noexcept { return rimpl_->rs_ipv4_addr_octets(); }

bool Ipv4Addr::is_unspecified() const noexcept { return rimpl_->rs_ipv4_addr_is_unspecified(); }

bool Ipv4Addr::is_loopback() const noexcept { return rimpl_->rs_ipv4_addr_is_loopback(); }

bool Ipv4Addr::is_private() const noexcept { return rimpl_->rs_ipv4_addr_is_private(); }

bool Ipv4Addr::is_link_local() const noexcept { return rimpl_->rs_ipv4_addr_is_link_local(); }

bool Ipv4Addr::is_multicast() const noexcept { return rimpl_->rs_ipv4_addr_is_multicast(); }

bool Ipv4Addr::is_broadcast() const noexcept { return rimpl_->rs_ipv4_addr_is_broadcast(); }

bool Ipv4Addr::is_documentation() const noexcept { return rimpl_->rs_ipv4_addr_is_documentation(); }

std::string Ipv4Addr::to_string() const noexcept { return std::string(rimpl_->rs_ipv4_addr_to_string()); }

bool Ipv4Addr::is_valid() const noexcept { return static_cast<bool>(rimpl_); }

} // namespace testcontainers

File: rust/src/system/ip/ipv4_addr.rs
use std::net::Ipv4Addr;

pub struct RsIpv4Addr {
    pub addr: Ipv4Addr,
}

pub fn rs_ipv4_addr_new(a: u8, b: u8, c: u8, d: u8) -> Box<RsIpv4Addr> {
    Box::new(RsIpv4Addr::new(Ipv4Addr::new(a, b, c, d)))
}

pub fn rs_ipv4_addr_localhost() -> Box<RsIpv4Addr> {
    Box::new(RsIpv4Addr::new(Ipv4Addr::LOCALHOST))
}

pub fn rs_ipv4_addr_unspecified() -> Box<RsIpv4Addr> {
    Box::new(RsIpv4Addr::new(Ipv4Addr::UNSPECIFIED))
}

pub fn rs_ipv4_addr_broadcast() -> Box<RsIpv4Addr> {
    Box::new(RsIpv4Addr::new(Ipv4Addr::BROADCAST))
}

pub fn rs_ipv4_addr_from_bits(bits: u32) -> Box<RsIpv4Addr> {
    Box::new(RsIpv4Addr::new(Ipv4Addr::from_bits(bits)))
}

pub fn rs_ipv4_addr_from_octets(octets: [u8; 4]) -> Box<RsIpv4Addr> {
    Box::new(RsIpv4Addr::new(Ipv4Addr::from(octets)))
}

pub fn rs_ipv4_addr_destroy(addr: Box<RsIpv4Addr>) {
    drop(addr);
}

impl RsIpv4Addr {
    pub fn new(addr: Ipv4Addr) -> Self {
        Self { addr }
    }

    pub fn rs_ipv4_addr_to_bits(&self) -> u32 {
        self.addr.to_bits()
    }

    pub fn rs_ipv4_addr_octets(&self) -> [u8; 4] {
        self.addr.octets()
    }

    pub fn rs_ipv4_addr_is_unspecified(&self) -> bool {
        self.addr.is_unspecified()
    }

    pub fn rs_ipv4_addr_is_loopback(&self) -> bool {
        self.addr.is_loopback()
    }

    pub fn rs_ipv4_addr_is_private(&self) -> bool {
        self.addr.is_private()
    }

    pub fn rs_ipv4_addr_is_link_local(&self) -> bool {
        self.addr.is_link_local()
    }

    // pub fn rs_ipv4_addr_is_global(&self) -> bool {
    //     self.addr.is_global()
    // }

    // pub fn rs_ipv4_addr_is_shared(&self) -> bool {
    //     self.addr.is_shared()
    // }

    // pub fn rs_ipv4_addr_is_benchmarking(&self) -> bool {
    //     self.addr.is_benchmarking()
    // }

    // pub fn rs_ipv4_addr_is_reserved(&self) -> bool {
    //     self.addr.is_reserved()
    // }

    pub fn rs_ipv4_addr_is_multicast(&self) -> bool {
        self.addr.is_multicast()
    }

    pub fn rs_ipv4_addr_is_broadcast(&self) -> bool {
        self.addr.is_broadcast()
    }

    pub fn rs_ipv4_addr_is_documentation(&self) -> bool {
        self.addr.is_documentation()
    }

    pub fn rs_ipv4_addr_to_string(&self) -> String {
        self.addr.to_string()
    }
}

